test/sqllogictest/literal_constraints.slt

```diff
@@ -729,6 +729,36 @@ WHERE
 ----
 2  l2  32
 
+# Implicit casts between literals and indexed column types shouldn't prevent index access
+
+statement ok
+CREATE TABLE foo(a SMALLINT, b INT, c BIGINT, v VARCHAR);
+
+statement ok
+CREATE DEFAULT INDEX ON foo;
+
+statement ok
+INSERT INTO foo VALUES (0, 1, 2, 'xxx'), (3, 4, 5, 'yyy');
+
+query T multiline
+EXPLAIN SELECT * FROM FOO
+WHERE a = 0 AND b = 1 AND c = 2 AND v = 'xxx'
+----
+Explained Query (fast path):
+  Project (#0..=#3)
+    ReadExistingIndex materialize.public.foo_primary_idx lookup value (0, 1, 2, "xxx")
+
+Used Indexes:
+  - materialize.public.foo_primary_idx
+
+EOF
+
+query IIIT rowsort
+SELECT * FROM FOO
+WHERE a = 0 AND b = 1 AND c = 2 AND v = 'xxx'
```

Comment:
Should we also add a test/support for the literal on the RHS?

Reply:
Yes, absolutely, thx! Done.

This case is handled by the canonicalization in `MirScalarExpr::reduce` around here:
```
                        } else if matches!(*func, BinaryFunc::Eq | BinaryFunc::NotEq
                            && expr2 < expr1
                        {
                            // Canonically order elements so that deduplication works better.
                            // Also, the below `Literal([c1, c2]) = record_create(e1, e2)` matching
                            // relies on this canonical ordering.
                            mem::swap(expr1, expr2);
                        }
```